### DISCUSSION TITLE
Csla.Core.UndoException:“Edit level mismatch in

### DISCUSSION THREAD

**Discussion by 455986780** [MarimerLLC/csla#2365] [Questions]
**Describe the bug**

 Under the main object, there is a collection of sub objects, which are bound to the grid control of windows forms.
When deleting rows from the grid, click save main object.ApplyEdit () error.

"Csla.Core.UndoException:“Edit level mismatch in ".

If the code under the main object is annotated, there will be no error.

        protected override void OnChildChanged(Csla.Core.ChildChangedEventArgs e)
        {
            /*
            if (e.ChildObject is SelQuerySqlCollection)
            {
                BusinessRules.CheckRules(SelQuerySqlsProperty);
                OnPropertyChanged(SelQuerySqlsProperty);
            }*/
            base.OnChildChanged(e);
        }

**Version and Platform**
CSLA version:  5.4.0
OS: Windows
Platform: WinForms

**Code that Fails**
     if (saveObject)
                {
                    if (business != null)
                        business.ApplyEdit();
 
![err](https://user-images.githubusercontent.com/58161541/102186335-6e156800-3eed-11eb-9a38-e4de47d6eb8c.jpg)

**Stack Trace or Exception Detail**
at Csla.Core.UndoableBase.AcceptChanges(Int32 parentEditLevel)\r\n   at Csla.Core.UndoableBase.Csla.Core.IUndoableObject.AcceptChanges(Int32 parentEditLevel, Boolean parentBindingEdit)\r\n   at Csla.BusinessListBase`2.AcceptChanges(Int32 parentEditLevel)\r\n   at Csla.BusinessListBase`2.Csla.Core.IUndoableObject.AcceptChanges(Int32 parentEditLevel, Boolean parentBindingEdit)\r\n   at Csla.Core.FieldManager.FieldDataManager.Csla.Core.IUndoableObject.AcceptChanges(Int32 parentEditLevel, Boolean parentBindingEdit)\r\n   at Csla.Core.UndoableBase.AcceptChanges(Int32 parentEditLevel)\r\n   at Csla.Core.BusinessBase.ApplyEdit()\r\n   at iPoint.iWork.Base.Win.Libary.Docenter.TableNameControl.RebindUI(Boolean saveObject, Boolean rebind) in D:\\TestProject\\UI\\WinForms\\TableNameControl.cs:line 955

**Additional context**

**Comment by rockfordlhotka** [comment]
I am assuming you are using the standard Microsoft datagrid control. Third party controls _hopefully_ follow all the data binding rules, but sometimes they don't.

Normally, when you bind a list to a datagrid, each time the user enters a row in the datagrid, the control will call `BeginEdit` on that child object (representing the row). When the user leaves the row normally the control calls `ApplyEdit`, and if the user presses ESC to leave the row the control calls `CancelEdit`.

All the rules about n-level undo and how it interacts with Windows Forms is documented in the [Using CSLA 3 book](https://store.lhotka.net/using-csla-net-version-30).

The most current example of Windows Forms binding with a datagrid is the [SimpleNTier sample](https://github.com/MarimerLLC/csla/tree/master/Samples/SimpleNTier).

**Comment by 455986780** [comment]
Thank you very much!

I refer to the object binding of "projecttracker". The sub object binding grid control is gridcontrol of devexpress.

Modifying and adding lines are normal when the root object is saved.

Occasionally, there is an error in deleting a row when the root object is saved.


     private void BindUI()
        {
            CurCatalogItem.BeginEdit();
            this.bsCatalog.DataSource = CurCatalogItem;
        }

    private bool RebindUI(bool saveObject, bool rebind)
        {
            this.Validate();
            this.ValidateChildren();
            // disable events
            this.bsCatalog.RaiseListChangedEvents = false;
            this.bsSubCatalogs.RaiseListChangedEvents = false;
            try
            {
                // unbind the UI
                UnbindBindingSource(this.bsSubCatalogs, saveObject, false);
                UnbindBindingSource(this.bsCatalog, saveObject, true);
                this.bsSubCatalogs.DataSource = this.bsCatalog;
                // save or cancel changes
                if (saveObject)
                {
                    this.CurCatalogItem.ApplyEdit();
                    try
                    {
                        CurCatalogItem = CurCatalogItem.Save();
                    }
                    catch (Csla.DataPortalException ex)
                    {
                        MessageBox.Show(ex.BusinessException.ToString(),
                          "Error saving", MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                        return false;
                    }
                    catch (System.Exception ex)
                    {
                        MessageBox.Show(ex.ToString(),
                          "Error Saving", MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
                        return false;
                    }
                }
                else
                    CurCatalogItem.CancelEdit();
                return true;
            }
            finally
            {
                // rebind UI if requested
                if (rebind)
                {
                    BindUI();
                }
                // restore events
                this.bsCatalog.RaiseListChangedEvents = true;
                this.bsSubCatalogs.RaiseListChangedEvents = true;
                if (rebind)
                {
                    // refresh the UI if rebinding
                    this.bsCatalog.ResetBindings(false);
                    this.bsSubCatalogs.ResetBindings(false);
                }
            }
        }

**Comment by 455986780** [comment]
I added an interface "frmorder" to "project Tracker" to reproduce errors.
After running, click the "delete grid row" button, and then click the "save" button to reproduce the error. If there is no error, add a new line on the grid, and then click the "delete grid row" button, and then click the "save" button to reproduce the error.

[ProjectTracker.zip](https://github.com/MarimerLLC/csla/files/5701314/ProjectTracker.zip)

**Comment by KC7465128305** [comment]
I have a WPF app using csla 5.4.0 and I have noticed that this error can be triggered if I put a breakpoint in the code and debug the app.  I learned to treat this as an inconvenience because so long as I don't put breakpoint then all is fine.  This doesn't happen all the time but it does happen from time to time.  It seems like when a breakpoint breaks it will in some circumstances trigger this EditLevel mismatch exception and this has nothing to do with BeginEdit, ApplyEdit, and CancelEdit matching.  This behavior has been going on in the earlier csla versions.

**Comment by pszittyay** [comment]
We have reproduced this error on a DevExpress Grid Control only when Filtering the DataSource.
We are also interested in helping adress this issue either by a PR or help testing

**Comment by rockfordlhotka** [comment]
@pszittyay if you have a solution, I'd love a PR.

The trick with these sorts of issue, is that the code _must work with the Microsoft datagrid_. Working with third party controls is fantastic, as long as the in-the-box controls continue to work.

It might be a problem with `SortedBindingList`, in which case identifying the issue with a unit test is best - skip data binding idiosyncrasies entirely. (though usually it is a data binding idiosyncrasy that causes the issue 😢)

**Comment by ashishsinha24** [comment]
> I added an interface "frmorder" to "project Tracker" to reproduce errors.
> After running, click the "delete grid row" button, and then click the "save" button to reproduce the error. If there is no error, add a new line on the grid, and then click the "delete grid row" button, and then click the "save" button to reproduce the error.
> 
> [ProjectTracker.zip](https://github.com/MarimerLLC/csla/files/5701314/ProjectTracker.zip)

Try replacing the delete code with this

`` private void btnRemoveItem_Click(object sender, EventArgs e)
``{
      ``     var id = (int)dataGridView1.SelectedRows[0].Cells["colId"].Value;``
       ``     CurOrder.OrderLines.Remove(id);``
``}``

**Comment by rockfordlhotka** [comment]
The most common cause of this, from years ago, was when a datagrid control didn't correctly interact with the `System.ComponentModel.IEditableObject` data binding interface on a per-row basis. 

What a datagrid control _must do_ is always invoke `BeginEdit`, and when the focus leaves the row it must _always_ call either `CancelEdit` or `EndEdit`.

Some datagrid controls didn't _always_ call the Cancel or End edit methods, depending on how focus left the datagrid for other controls in the UI. That would leave the row on which the user was last editing in the middle of an edit, because neither Cancel nor End was invoked.

**Comment by ashishsinha24** [comment]
I stopped relying on the grid's events (At least for the Delete Operation) and instead manually call the parent object's remove function.

**Comment by rockfordlhotka** [comment]
Did that get your app working?

**Comment by ashishsinha24** [comment]
> Did that get your app working?

O Yes. Way back in 2006-2007, I faced the same problem in CSLA 3.7 & DevExpress Grid. 
That's when I switched to this workaround of directly interacting with the Object in Delete Operations. The New and Update functions worked fine.

**Comment by rockfordlhotka** [comment]
This is always the hard part - I do (did) all my testing against the Microsoft controls, because there's no way to test against all the other random controls out there. Sometimes a workaround is all there is...

@455986780 I'm not sure if @ashishsinha24 's workaround will solve your issue?

**Comment by ashishsinha24** [comment]
Let me know if this works on your end.

[ProjectTracker.zip](https://github.com/MarimerLLC/csla/files/6787358/ProjectTracker.zip)

**Comment by qhliao66** [comment]
I disabled "DisableIEditableObject= true" in the business object constructor,

It's running well now!

**Comment by michaelcsikos** [comment]
Our grids are based on the standard `DataGridView`. The `EditLevelMismatch` exception is easy to reproduce when sorting or filtering a `BusinessBindingListBase`. As Rocky said, the `EndEdit()` doesn't always get called when the sort or filter is applied. To solve this, you can make a copy of the `SortedBindingList` and `FilteredBindingList` classes, make the `ApplySort` and `ApplyFilter` methods `vrtual`, then inherit these classes and add the following overrides:

``` cs
public override void ApplyFilter(PropertyDescriptor property, object filter)
{
    OnFilterChanging();

    base.ApplyFilter(property, filter);

    OnFilterChanged();
}

public override void RemoveFilter()
{
    OnFilterChanging();

    base.RemoveFilter();

    OnFilterChanged();
}

public event EventHandler FilterChanging;

private void OnFilterChanging()
{
    if (FilterChanging != null)
        FilterChanging(this, EventArgs.Empty);
}

public event EventHandler FilterChanged;

private void OnFilterChanged()
{
    if (FilterChanged != null)
        FilterChanged(this, EventArgs.Empty);
}

public override void ApplySort(PropertyDescriptor property, ListSortDirection direction)
{
    OnSortChanging();

    base.ApplySort(property, direction);

    OnSortChanged();
}

public override void ApplySort(string propertyName, ListSortDirection direction)
{
    OnSortChanging();

    base.ApplySort(propertyName, direction);

    OnSortChanged();
}

public void RefreshSort()
{
    if (IsSorted)
        ApplySort(SortProperty, SortDirection);
}

public override void RemoveSort()
{
    OnSortChanging();

    base.RemoveSort();

    OnSortChanged();
}

public event EventHandler SortChanging;

private void OnSortChanging()
{
    if (SortChanging != null)
        SortChanging(this, EventArgs.Empty);
}

public event EventHandler SortChanged;

private void OnSortChanged()
{
    if (SortChanged != null)
        SortChanged(this, EventArgs.Empty);
}
```

The `FilterChanged` and `SortChanged` events are not essential to solve this, but may be useful.

Then in your UI, subscribe the appropriate events if sorting and/or filtering is supported:

``` cs
private void BindUI()
{
    // ...
    filteredBindingList.SortChanging   += OnFilterOrSortChanging;
    filteredBindingList.FilterChanging += OnFilterOrSortChanging;
    // ...
}

private void OnFilterOrSortChanging(object sender, EventArgs e)
{
    bindList.EndEdit();
}
```

**Reply by pszittyay** [reply]
@michaelcsikos you mean override the SortedBindingList and FilteredBindingList in CSLA namespace? I can make a PR based on your solution an test it on the DataGridView and the DevExpress GridView.  What do you think @rockfordlhotka ?

**Reply by michaelcsikos** [reply]
@pszittyay, it doesn't have to be in the CSLA namespace for you to test it. From memory, `SortedBindingList` and `FilteredBindingList` don't depend on anything else in CSLA, so you can do a copy and paste to your own project, then add `virtual` to the required methods. We did this in our own project so that we could keep the changes to the original CSLA classes to a minimum, which makes it easier to update should anything change in the future. (We have since modified our copies of the classes to allow sorting or filtering on multiple `PropertyDescriptor` properties simultaneously instead of just one.)

**Reply by pszittyay** [reply]
@michaelcsikos thanks for the response and you have great memory, both classes don't depend on anything else from de CSLA namespace. 
Reading again your example, i fail to understand is how this 2 collections (filteredBindingList  and bindList) work together. I assume that bindList is of type BusinessBindingListBase which i bound to my grid via static method BindingSourceHelper.InitializeBindingSourceTree() and the Filter/Sort is done automatically by the grid.

**Reply by michaelcsikos** [reply]
@pszittyay Sorry, I should have specified `bindList` is a standard WinForms `BindingSource` which would then be the `DataSource` of the `DataGridView`. The sorted and filtered binding lists are usually composed or chained together, so the `BusinessBindingListBase` is wrapped in a `SortedBindingList`, which is then wrapped in a `FilteredBindingList`, which is then set as the `DataSource` for the `BindingSource`, which is finally the `DataSource` for the grid.

**Reply by pszittyay** [reply]
@michaelcsikos thanks again for taking the time to response. I will test this and get back to this discussion with the result. It also got me thinking how to make it simple and avoid this boilerplate code that i would have to repeat in all grids bound to a BusinessBindingListBase

**Comment by rockfordlhotka** [comment]
I did just fix a bug with `BusinessBindingListBase` in terms of how it raised some `ListChanged` events. This _might_ help resolve some of the issues being discussed in this thread?

The fix is in the new v5.5.0 prerelease I just pushed to NuGet.

**Reply by pszittyay** [reply]
@rockfordlhotka i can confirm that the problem is resolve in version 5.5.0.
Thanks very much for the effort, i would like to invite you and @michaelcsikos with a beer from the distance (I am from Argentina.)

**Reply by michaelcsikos** [reply]
Cheers! (I'm in Australia. It will be beer o'clock in a few hours.)
